File: bubus-ts/examples/forwarding_between_busses.ts
#!/usr/bin/env -S node --import tsx
// Run: node --import tsx examples/forwarding_between_busses.ts

import { z } from 'zod'

import { BaseEvent, EventBus } from '../src/index.js'

const ForwardedEvent = BaseEvent.extend('ForwardedEvent', {
  message: z.string(),
})

async function main(): Promise<void> {
  const busA = new EventBus('BusA')
  const busB = new EventBus('BusB')
  const busC = new EventBus('BusC')

  const handleCounts = {
    BusA: 0,
    BusB: 0,
    BusC: 0,
  }

  const seenEventIds = {
    BusA: new Set<string>(),
    BusB: new Set<string>(),
    BusC: new Set<string>(),
  }

  // Each bus handles the typed event locally.
  // In a forwarding cycle, loop prevention should keep each bus to one handle.
  busA.on(ForwardedEvent, (event) => {
    handleCounts.BusA += 1
    seenEventIds.BusA.add(event.event_id)
    console.log(`[BusA] handled ${event.event_id} (count=${handleCounts.BusA})`)
  })

  busB.on(ForwardedEvent, (event) => {
    handleCounts.BusB += 1
    seenEventIds.BusB.add(event.event_id)
    console.log(`[BusB] handled ${event.event_id} (count=${handleCounts.BusB})`)
  })

  busC.on(ForwardedEvent, (event) => {
    handleCounts.BusC += 1
    seenEventIds.BusC.add(event.event_id)
    console.log(`[BusC] handled ${event.event_id} (count=${handleCounts.BusC})`)
  })

  // Forward all events in a ring:
  // A -> B -> C -> A
  // Expected for one dispatch from A: event path becomes [A, B, C] and stops.
  // The C -> A edge is skipped because A is already in event_path.
  busA.on('*', busB.emit)
  busB.on('*', busC.dispatch)
  busC.on('*', busA.dispatch)

  console.log('Dispatching ForwardedEvent on BusA with cyclic forwarding A -> B -> C -> A')

  const event = busA.dispatch(
    ForwardedEvent({
      message: 'hello across 3 buses',
    })
  )

  // done() waits for handlers on all forwarded buses, not just the origin bus.
  await event.done()
  await Promise.all([busA.waitUntilIdle(), busB.waitUntilIdle(), busC.waitUntilIdle()])

  const path = event.event_path
  const totalHandles = handleCounts.BusA + handleCounts.BusB + handleCounts.BusC

  console.log('\nFinal propagation summary:')
  console.log(`- event_id: ${event.event_id}`)
  console.log(`- event_path: ${path.join(' -> ')}`)
  console.log(`- handle counts: ${JSON.stringify(handleCounts)}`)
  console.log(`- unique ids seen per bus: A=${seenEventIds.BusA.size}, B=${seenEventIds.BusB.size}, C=${seenEventIds.BusC.size}`)
  console.log(`- total handles: ${totalHandles}`)

  const handledOncePerBus = handleCounts.BusA === 1 && handleCounts.BusB === 1 && handleCounts.BusC === 1
  const visitedThreeBuses = path.length === 3

  if (handledOncePerBus && visitedThreeBuses) {
    console.log('\nLoop prevention confirmed: each bus handled the event at most once.')
  } else {
    console.log('\nUnexpected forwarding result. Check handlers/forwarding setup.')
  }

  console.log('\n=== BusA logTree() ===')
  console.log(busA.logTree())
  console.log('\n=== BusB logTree() ===')
  console.log(busB.logTree())
  console.log('\n=== BusC logTree() ===')
  console.log(busC.logTree())
}

await main()

File: examples/forwarding_between_busses.py
#!/usr/bin/env -S uv run python
"""Run: uv run python examples/forwarding_between_busses.py"""

import asyncio

from bubus import BaseEvent, EventBus


class ForwardedEvent(BaseEvent[None]):
    message: str


async def main() -> None:
    bus_a = EventBus('BusA')
    bus_b = EventBus('BusB')
    bus_c = EventBus('BusC')

    try:
        handle_counts = {'BusA': 0, 'BusB': 0, 'BusC': 0}
        seen_event_ids = {'BusA': set[str](), 'BusB': set[str](), 'BusC': set[str]()}

        def on_a(event: ForwardedEvent) -> None:
            handle_counts['BusA'] += 1
            seen_event_ids['BusA'].add(event.event_id)
            print(f'[BusA] handled {event.event_id} (count={handle_counts["BusA"]})')

        def on_b(event: ForwardedEvent) -> None:
            handle_counts['BusB'] += 1
            seen_event_ids['BusB'].add(event.event_id)
            print(f'[BusB] handled {event.event_id} (count={handle_counts["BusB"]})')

        def on_c(event: ForwardedEvent) -> None:
            handle_counts['BusC'] += 1
            seen_event_ids['BusC'].add(event.event_id)
            print(f'[BusC] handled {event.event_id} (count={handle_counts["BusC"]})')

        bus_a.on(ForwardedEvent, on_a)
        bus_b.on(ForwardedEvent, on_b)
        bus_c.on(ForwardedEvent, on_c)

        # Ring forwarding:
        # A -> B -> C -> A
        bus_a.on('*', bus_b.emit)
        bus_b.on('*', bus_c.emit)
        bus_c.on('*', bus_a.emit)

        print('Dispatching ForwardedEvent on BusA with cyclic forwarding A -> B -> C -> A')

        event = bus_a.emit(ForwardedEvent(message='hello across 3 buses'))
        await event
        await asyncio.gather(bus_a.wait_until_idle(), bus_b.wait_until_idle(), bus_c.wait_until_idle())

        path = event.event_path
        total_handles = handle_counts['BusA'] + handle_counts['BusB'] + handle_counts['BusC']

        print('\nFinal propagation summary:')
        print(f'- event_id: {event.event_id}')
        print(f'- event_path: {" -> ".join(path)}')
        print(f'- handle counts: {handle_counts}')
        print(
            '- unique ids seen per bus: '
            f'A={len(seen_event_ids["BusA"])}, '
            f'B={len(seen_event_ids["BusB"])}, '
            f'C={len(seen_event_ids["BusC"])}'
        )
        print(f'- total handles: {total_handles}')

        handled_once_per_bus = handle_counts['BusA'] == 1 and handle_counts['BusB'] == 1 and handle_counts['BusC'] == 1
        visited_three_buses = len(path) == 3

        if handled_once_per_bus and visited_three_buses:
            print('\nLoop prevention confirmed: each bus handled the event at most once.')
        else:
            print('\nUnexpected forwarding result. Check handlers/forwarding setup.')

        print('\n=== BusA log_tree() ===')
        print(bus_a.log_tree())
        print('\n=== BusB log_tree() ===')
        print(bus_b.log_tree())
        print('\n=== BusC log_tree() ===')
        print(bus_c.log_tree())
    finally:
        await bus_a.stop(clear=True, timeout=0)
        await bus_b.stop(clear=True, timeout=0)
        await bus_c.stop(clear=True, timeout=0)


if __name__ == '__main__':
    asyncio.run(main())
